managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/PositionImpl.java

```diff
@@ -102,20 +99,27 @@ public PositionImpl getNext() {
      */
     @Override
     public String toString() {
-        return String.format("%d:%d", ledgerId, entryId);
+        return ledgerId + ":" + entryId;
     }
 
     @Override
-    public int compareTo(PositionImpl other) {
-        checkNotNull(other);
+    public int compareTo(PositionImpl that) {
+        if (this.ledgerId != that.ledgerId) {
+            return (this.ledgerId < that.ledgerId ? -1 : 1);
+        }
+
+        if (this.entryId != that.entryId) {
+            return (this.entryId < that.entryId ? -1 : 1);
+        }
 
-        return ComparisonChain.start().compare(this.ledgerId, other.ledgerId).compare(this.entryId, other.entryId)
-                .result();
+        return 0;
     }
 
     @Override
     public int hashCode() {
-        return Objects.hashCode(ledgerId, entryId);
+        int result = (int) (ledgerId ^ (ledgerId >>> 32));
+        result = 31 * result + (int) (entryId ^ (entryId >>> 32));
+        return result;
```

Comment:
I am curious if we drop `ackSet` from hashCode on purpose or not.

Reply:
I intentionally didn't add that since this PR is a refactoring. Surprising things could happen if it would be added as part of the hash code. It's not a very good design to have the ackSet as part of the PositionImpl in the first place. The mutable design of PositionImpl is also questionable.

Reply:
> Surprising things could happen if it would be added as part of the hash code.

I looked into this a bit. It turns out that it can't be added into `hashcode`. Related method is 
`org.apache.pulsar.broker.transaction.pendingack.impl.PendingAckHandleImpl#syncBatchPositionAckSetForTransaction`.